### ISSUE TITLE
proposal: support error formatting with the %+v directive

### ISSUE THREAD

**Issue by stephenh** [DataDog/dd-trace-go#445]
Currently when we call `setTag(ext.Error, err)` the stack trace is from the place that called `setTag`.

This is sometimes useful, but we almost always want the stack trace from where the `err` originated.

Which granted, is a common pain point in Go; we've taken to using the pkg/errors project (https://github.com/pkg/errors), which means if we use its error wrapper, we can do things like:

```go
span.SetTag(ext.Error, err)
if _, ok := err.(fmt.Formatter); ok {
  span.SetTag(ext.ErrorStack, fmt.Sprintf("%+v", err))
}
```

And get an actually useful stack trace.

Granted, dd-trace-go shouldn't assume programs specifically use pkg/errors, but pkg/errors's approach (as you can see in the snippet) is that it implements `fmt.Formatter`, and so if you format with the `%+v`, you'll get back a stack trace.

I have no idea if other error-wrapping libraries follow a similar approach, but it would be pretty great if they did.

Do you know if there is a psuedo-standard within go that dd-trace-go could probe support for, i.e. with the `fmt.Formatter` check, and then if so, use the err's incoming stack trace?

**Comment by gbbr** [comment]
If I understand correctly, you are proposing a change where we would check inside `dd-trace-go` if the error implements `fmt.Formatter` and use `fmt.Sprintf("%+v", err)` as the value of the error stack tag. Is that correct?

If so, I'm not sure if it's common, outside of `pkg/errors`, for errors to implement `fmt.Formatter`. Furthermore, since it's so easy to do this (as you've proven yourself), I don't see any reason to have this in the library.

Either way, happy to discuss further.

**Comment by stephenh** [comment]
> Is that correct?

Yes.

> Furthermore, since it's so easy to do this (as you've proven yourself)

You are right, albeit that's b/c right now we're not using any of the out-of-the-box dd-trace-go instrumentations, we just have ~a few custom implementations that manually create spans/etc.

So, yes, for now we've copy/pasted this snippet around into ~2-3 places, and sent an FYI in our eng slack channel of "btw, if you're calling `setTag(error)` please include these three extra lines as well". So that's fine and we'll keep doing that.

Just would be nice to not have to remember/duplicate that bookkeeping.

I agree it would be a great/more-obvious addition if this was a widely-adopted golang idiom; perhaps as a maintainer of a package/service where that would be useful for literally all of your users, you could drive that as a useful thing for the ecosystem to agree/standardize on. :-) (I'm being admittedly selfish in not volunteering to do that personally.)

Also, fwiw/in terms of "I don't see any reason to have this in the library.", I'll just relate my personal experience as a ~golang newbie, that when I first looked at the DD spans for a bug I was triaging, I was pretty disappointed/frustrated that it was not the real cause, and the stacktrace-at-point-of-setTag was basically useless, for what I needed at the time.

And so that fed into the "error handling in go sucks" feeling/myth. I mean, yeah, a lot of that is on us to learn what go idioms we want to adopt to work around that, but it'd also be nice if it worked out-of-the-box (granted, this is getting into religious territory, about whether go's built-in errors should have stack traces attached to them natively, vs. hacked on the side by various disparate community libraries like pkg/errors, go-errors, palantir/stacktrace, merry, etc. And so, yes, very much outside of the purvey of dd-trace-go per se, but just relating my N=1 experience report that its indirectly affecting your users and making DD traces/spans from golang apps less useful than they could be.)

Feel free to close/triage as you like. :-)

**Comment by gbbr** [comment]
> So, yes, for now we've copy/pasted this snippet around into ~2-3 places, and sent an FYI in our eng slack channel of "btw, if you're calling setTag(error) please include these three extra lines as well". So that's fine and we'll keep doing that.

The best thing I can suggest in this case is to maybe create a package that does this in your repository, something like `traceutil`, rather than copy/pasting everywhere.

> I was pretty disappointed/frustrated that it was not the real cause, and the stacktrace-at-point-of-setTag was basically useless, for what I needed at the time.

I think this is a very good point, and maybe we can extract something useful from it as a change in the project. What do you think could make it better? It would be interesting if you could provide an example of what the stack trace looked like, and what you would prefer it to look like.

**Comment by stephenh** [comment]
> something like traceutil

Yeah, that's true, we'll do something like that.

> if you could provide an example of what the stack trace looked like

Sure; these are pretty tame stack traces, so I'll just paste the real ones; here was the original stacktrace-at-point-of-setTag:

```
goroutine 44178808 [running]:
runtime/debug.Stack(0x7b2400, 0xc421a5dec0, 0x826def)
	/usr/local/go/src/runtime/debug/stack.go:24 +0xa7
github.com/remind101/r101-permissions/vendor/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.(*span).setTagError(0xc424f43720, 0x7cace0, 0xc4202ace40)
	/go/src/github.com/remind101/r101-permissions/vendor/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer/span.go:117 +0x204
github.com/remind101/r101-permissions/vendor/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.(*span).SetTag(0xc424f43720, 0x82573b, 0x5, 0x7cace0, 0xc4202ace40)
	/go/src/github.com/remind101/r101-permissions/vendor/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer/span.go:84 +0x36e
github.com/remind101/r101-permissions/vendor/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/opentracer.(*span).SetTag(0xc4202acc80, 0x82573b, 0x5, 0x7cace0, 0xc4202ace40, 0xc422789750, 0xc427527fc8)
	/go/src/github.com/remind101/r101-permissions/vendor/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/opentracer/span.go:81 +0x5b
github.com/remind101/r101-permissions/rules.ResolveAllChecks.func1(0x7fc634c380f0, 0xc4284d2030, 0x876120, 0xc4202acc80, 0xc4209e1110, 0xc4280b9d40, 0xc4202acd40)
	/go/src/github.com/remind101/r101-permissions/rules/resolve.go:92 +0x101
created by github.com/remind101/r101-permissions/rules.ResolveAllChecks
	/go/src/github.com/remind101/r101-permissions/rules/resolve.go:78 +0x1cb
```

Where this `ResolveAllChecks` was in charge of the creating/finishing a span around its business logic, and its business logic was resolving N rules in parallel. If any of the rules errors, it bubbles that error up into the `setTag(error)` call.

So, with this, we just knew, "well, _some_ rule failed".

Thankfully our code was already using the pkg/errors wrapping library/idiom, such that when the individual rule failed, it was already passing back an error-that-captured-its-stacktrace, and so with just making the formatter change, we instead got this stack trace:

```
runtime error: invalid memory address or nil pointer dereference
Panic in evalPredicate
github.com/remind101/r101-permissions/rules.evalPredicate.func1
	/go/src/github.com/remind101/r101-permissions/rules/operators.go:13
runtime.call32
	/usr/local/go/src/runtime/asm_amd64.s:573
runtime.gopanic
	/usr/local/go/src/runtime/panic.go:502
runtime.panicmem
	/usr/local/go/src/runtime/panic.go:63
runtime.sigpanic
	/usr/local/go/src/runtime/signal_unix.go:388
github.com/remind101/r101-permissions/store.GetSchoolsInADistrict
	/go/src/github.com/remind101/r101-permissions/store/organizations_store.go:114
github.com/remind101/r101-permissions/rules.IsFullyPaidOrganization.func1
	/go/src/github.com/remind101/r101-permissions/rules/organization.go:176
github.com/remind101/r101-permissions/rules.evalPredicate
	/go/src/github.com/remind101/r101-permissions/rules/operators.go:18
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:2361
```

Which was that "ooohhh, `GetSchoolsInADistrict` is failing" hint that we needed to move on to the next stage of debugging.

Thinking about this more concretely, I guess without the idiom of pkg/errors-wrapped-errors + use-pkg-errors-aware-setTag, we'd want something like "create a span for every ...function/gorountine... that does a defer/recover", because those are generally are our smallest slice points for "recover random failures within this chunk of logic" that are helpful to pinpoint when debugging. And so then the spans would be closer to the point of failure, vs. where it was now in a higher-level abstraction.

I'm not entirely sure if we want that many spans; its kind of a balance, more spans are helpful for debugging, but we can't have a span per stack frame. :-)

So, not sure. Thanks for at least thinking about this; again I realize I'm asking you to help with what is generally just a known-gotcha with error handling in go, so at the end of the day it may still just be on each user of dd-trace-go to bring their own "we do/do not use userland-library-x for wrapped errors" to the table.

Maybe that could be a config option? Like let users pass in an error formatter when configuring the tracer that is aware of their own particular userland-library?

I've seen something similar in pino, a logging library from the JS/TS world, where if you use "smarter" error logging libraries (JS doesn't natively have "caused by" errors like the JVM world has, and so some userland libraries shim this in), you can tell it to use your formatter on the way to the wire/stdout.

**Comment by gbbr** [comment]
I understand. Thanks for posting both examples, that was very helpful. This is definitely something we can improve on.

My concern is that if we start making the assumption that if the error implements `fmt.Formatter` and it is printed with the `%+v` directive, it will pretty-print its stack, that'd be a bad assumption to make.

Nevertheless, I feel that the main issue here is that the stack starts at the call of the internal `setTagError`, not necessarily that it's not pretty printed. I would like to suggest the following:

* Instead of calling `debug.Stack` [here](https://github.com/DataDog/dd-trace-go/blob/30ce4e46d1fe28b43d2cfd8347ec1e7144575f06/ddtrace/tracer/span.go#L136), simply use the newly added (in #436) function `takeStackTrace` to skip the frames that are of no interest. This will make the stack start in your code.
* Update the documentation accordingly.
* Benchmark before and after to be aware of the impact.

Do you think this solution would be along the lines of what you're looking for?

**Comment by gbbr** [comment]
It seems that the authors of #443 intend to contribute a new global tracer option, which will allow you to skip a certain number of stack frames which are not of interest. Let's wait for that to happen, it might solve this problem too. Let me know what you think.

**Comment by stephenh** [comment]
>  stack starts at the call of the internal setTagError, not necessarily that it's not pretty printed

Yes, agreed the pretty printing itself is not important.

> a certain number of stack frames which are not of interest

This is really a separate issue.

We don't care about skipping a few frames of the `setTag`/`takeStackTrace` stacktrace that are the internals of the ddtrace go impl (although I get that is nice from a UX perspective).

The issue is that the stacktrace we want has frames that are from a completely different point in the codebase.

I.e. the real bug in our code happened in a goroutine, which we captured that stacktrace at-the-time-it-happened, with (say) A/B/C methods in the stackframes, but by the time we percolate the error up to the `span.setTag` location of our codebase, the A/B/C methods would just not even exist in any stacktrack/`takeStackTrace` that we would capture at the `setTag` location.

So, using the `fmt.Formatter` `+v` was not about making the `setTag`-based-trace look better, it has about the `pkg/errors` `fmt.Formatter` impl realizing "oh, you've got a _completely different stack trace_ in this error that I will also include in the output for you".

**Comment by gbbr** [comment]
Ah, right. Sorry. My bad. I totally get what you're saying now.

**Comment by gbbr** [comment]
I think this would be a great addition. I'm just a bit skeptical that it's fine to make the assumption "_if the error implements `fmt.Formatter`, then it will print a useful stack given the `+v` formatter_". I would definitely like us to benefit from this functionality, I just don't have a good idea right now. Open to suggestions.

**Comment by gbbr** [comment]
@cgilmour do you have any thoughts on the above?

**Comment by stephenh** [comment]
@gbbr yeah, I agree the `fmt.Formatter` type check is not an acceptable heuristic. Seems like ideally they're be a `Stacktraceable` type in the core golang library, but you could also introduce your own, and then somehow let the user, in the ddtrace config, say "I'm using userland error libraries x/y/z, so please mixin our `NestedTrace`/whatever implementation into their types". ...I'm really at the boundaries of my current golang knowledge though (i.e. I don't really know how go does the internal bookkeeping of what types a struct implements for the `foo.(some.Type)` type checks), so defer to you guys. :-)

**Comment by gbbr** [comment]
Ok, I have an idea. I suggest adding another tag, called `error.detail`, which will be populated when the error implements either:
* https://godoc.org/golang.org/x/xerrors#Formatter
* https://godoc.org/fmt#Formatter

This is in accordance with the [planned error proposal](https://go.googlesource.com/proposal/+/master/design/29934-error-values.md) for Go 1.13 and `%+v` [seems to be the standard verb](https://github.com/golang/go/wiki/ErrorValueFAQ#what-formatting-verb-should-i-use-to-display-errors) for displaying error details.

**Comment by stephenh** [comment]
Wow, that was serendipitous! :-) Thanks @gbbr!
